lib/index.ts

```diff
@@ -1,317 +1,557 @@
-        enumerable: true,
-        value: unbind
-    });
-
-    Object.defineProperty(this, 'getBoundData', {
-        enumerable: true,
-        value: getBoundData
-    });
-
-    Object.defineProperty(this, 'enumerateLoadedClasses', {
-        enumerable: true,
-        value: enumerateLoadedClasses
-    });
-
-    Object.defineProperty(this, 'enumerateLoadedClassesSync', {
-        enumerable: true,
-        value: enumerateLoadedClassesSync
-    });
-
-    Object.defineProperty(this, 'choose', {
-        enumerable: true,
-        value: choose
-    });
-
-    Object.defineProperty(this, 'chooseSync', {
-        enumerable: true,
-        value: function (specifier) {
-            const instances = [];
-            choose(specifier, {
-                onMatch: function (i) {
-                    instances.push(i);
-                },
-                onComplete: function () {
-                }
-            });
-            return instances;
-        }
-    });
+    export const api = <Cache<NativeFunction>> getApi();
 
-    this.schedule = function (queue, work) {
-        const id = ptr(nextId++);
-        scheduledWork[id.toString()] = work;
+    export const available = api !== null;
 
-        if (workCallback === null) {
-            workCallback = new NativeCallback(performScheduledWorkItem, 'void', ['pointer']);
-        }
+    export const classes = available ? classRegistry : {};
 
-        Script.pin();
-        api.dispatch_async_f(queue, id, workCallback);
-    };
+    export const protocols = available ? protocolRegistry : {};
 
-    function performScheduledWorkItem(rawId) {
-        const id = rawId.toString();
-        const work = scheduledWork[id];
-        delete scheduledWork[id];
+    export const Object = <ObjectConstructor> (<any> ObjCObject);
 
-        if (NSAutoreleasePool === null)
-            NSAutoreleasePool = classRegistry.NSAutoreleasePool;
+    export const Protocol = <ProtocolConstructor> (<any> ObjCProtocol);
 
-        const pool = NSAutoreleasePool.alloc().init();
-        let pendingException = null;
-        try {
-            work();
-        } catch (e) {
-            pendingException = e;
-        }
-        pool.release();
+    export const Block = <BlockConstructor> (<any> ObjCBlock);
 
-        setImmediate(performScheduledWorkCleanup, pendingException);
-    }
+    export const mainQueue = api._dispatch_main_q;
 
-    function performScheduledWorkCleanup(pendingException) {
-        Script.unpin();
+    export function registerProxy(properties: ObjC.ProxySpec) {
+        type ProxyMethod = ObjC.UserMethodImplementation<any, any, any>
+            | ObjC.MethodSpec<ObjC.UserMethodImplementation<any, any, any>>;
 
-        if (pendingException !== null) {
-            throw pendingException;
-        }
-    }
+        const protocols = properties.protocols || [];
+        const methods = properties.methods || {};
+        const events = properties.events || {};
```

Comment:
Let's use the TypeScript 3.7 nullish coalescing operator feature - `??` – for such situations. Though in this case it would be better to use destructuring:

```
        const {
            protocols = [],
            methods = {},
            events = {},
        } = properties;
```